src/HtmlReader/lib.ts

```diff
@@ -0,0 +1,346 @@
+  currentPage: number;
+  currentPageFloor: number;
+  isAtEnd: boolean;
+  isAtStart: boolean;
+};
+
+/**
+ * Calculate current position information
+ */
+export function calcPosition(
+  iframe: HTMLIFrameElement,
+  isScrolling: boolean
+): CalculatedPosition {
+  /**
+   * Calculate the current scroll position and number of pages,
+   * and thus the page index, total pages and progression
+   */
+  const containerHeight = iframe.offsetHeight;
+  const containerWidth = iframe.offsetWidth;
+  const isHorizontalPaginated = !isScrolling;
+  const containerSize = isHorizontalPaginated
+    ? containerWidth
+    : containerHeight;
+  const pageSize = containerSize;
+  const html = getIframeHTML(iframe);
+  const resourceHeight = html.scrollHeight;
+  const resourceWidth = html.scrollWidth;
+  const resourceSize = isHorizontalPaginated ? resourceWidth : resourceHeight;
+  const { x: scrollXPosition, y: scrollYPosition } = getScrollPosition(html);
+  const scrollPosition = isHorizontalPaginated
+    ? scrollXPosition
+    : scrollYPosition;
+  const totalPages =
+    Math.ceil((resourceSize - containerSize) / containerSize) + 1;
+  const progression = scrollPosition / resourceSize;
+
+  const isAtBottom = scrollPosition + containerSize === resourceSize;
+
+  // we use round to get the closest page to the scrollTop, unless
+  // you are at the complete bottom.
+  const currentPage = isAtBottom
+    ? totalPages
+    : Math.round(progression * totalPages) + 1;
+  // we use floor to get the nearest fully read page.
+  const currentPageFloor = Math.floor(progression * totalPages) + 1;
+
+  // you're at the end if the scroll position + containerSize === resourceSize
+  const isAtEnd = currentPage === totalPages;
+  const isAtStart = currentPage === 1;
+
+  return {
+    isHorizontalPaginated,
+    containerSize,
+    containerWidth,
+    containerHeight,
+    pageSize,
+    resourceSize,
+    scrollPosition,
+    totalPages,
+    progression,
+    currentPage,
+    currentPageFloor,
+    isAtEnd,
+    isAtStart,
+  };
+}
+
+/**
+ * Gets scroll position of an element
+ */
+export function getScrollPosition(
+  element: HTMLElement | undefined | null
+): { x: number; y: number } {
+  if (!element) return { x: 0, y: 0 };
+  const { scrollTop, scrollLeft } = element;
+  return {
+    x: scrollLeft,
+    y: scrollTop,
+  };
+}
+
+/**
+ * Fetches a resource as text
+ */
+export async function fetchAsTxt(url: string): Promise<string> {
+  const res = await fetch(url);
+  const txt = await res.text();
+  return txt;
+}
+
+/**
+ * Inject some raw JS into the iframe document
+ */
+export function injectJS(document: Document): void {
+  const script = document.createElement('script');
+  script.setAttribute('type', 'text/javascript');
+  script.textContent = `
+    window.onload = () => {
+      window.top.postMessage({source: 'reader-iframe', type: 'IFRAME_LOADED'})
+    }
+  `;
+  document.head.appendChild(script);
+}
+
+export function getMaybeIframeHtml(
```

Comment:
dumb question - there truly isn't a better way to do an optional?

Reply:
Is this so that you don't have to check for `undefined` every time you use `getIframeHtml`?

Reply:
Exactly. I wanted one function for when it _has_ to be defined, and another for when maybe it's defined and maybe not.